File: python/101-110/Binary Tree Zigzag Level Order Traversal.py
    # @param root, a tree node
    # @return a list of lists of integers
    def zigzagLevelOrder(self, root):
        if root ==None:
            return []
        result=[]
        upper,down=[],[]
        upper.append(root)
        toLeft=True
        while len(upper)!=0:
            temp=[]
            i=len(upper)-1
            while i>=0:
                node=upper[i]
                temp.append(node.val)
                if toLeft:
                    if node.left!=None:
                        down.append(node.left)
                    if node.right!=None:
                        down.append(node.right)
                else:
                    if node.right!=None:
                        down.append(node.right)
                    if node.left!=None:
                        down.append(node.left)
                i-=1
            toLeft=not toLeft
            result+=[temp]
            upper=[]
            if len(down)>0:
                upper=down[:]
                down=[]
        return result

File: Cpp/101-110/Binary Tree Zigzag Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 
 /*
    runtime: 4ms
    error: 0
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int> > res;
        if(root==NULL) return res;
        stack<TreeNode*> myStack;
        stack<TreeNode*> temp;
        int isleft=1;
        myStack.push(root);
        while(!myStack.empty()){
            vector<int> one;
            while(!myStack.empty()){
                TreeNode* node = myStack.top();
                myStack.pop();
                one.push_back(node->val);
                if(isleft==1){
                    if(node->left!=NULL)  temp.push(node->left);
                    if(node->right!=NULL)  temp.push(node->right);
                }else{
                    if(node->right!=NULL) temp.push(node->right);
                    if(node->left!=NULL)  temp.push(node->left);
                }
            }
            isleft*=-1;
            myStack.swap(temp);
            res.push_back(one);
        }
        return res;
    }
};
